pkg/apis/servicecatalog/validation/instance.go

```diff
@@ -335,35 +337,63 @@ func ValidateServiceInstanceReferencesUpdate(new *sc.ServiceInstance, old *sc.Se
 func validatePlanReference(p *sc.PlanReference, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
+	// helper function to test that exactly one set of plan references are set
+	b2i := func(b bool) int8 {
+		if b {
+			return 1
+		}
+		return 0
+	}
+
 	// Just to make reading of the conditionals in the code easier.
-	externalClassSet := p.ClusterServiceClassExternalName != ""
-	externalPlanSet := p.ClusterServicePlanExternalName != ""
+	externalClassNameSet := p.ClusterServiceClassExternalName != ""
+	externalPlanNameSet := p.ClusterServicePlanExternalName != ""
+	externalClassIDSet := p.ClusterServiceClassExternalID != ""
+	externalPlanIDSet := p.ClusterServicePlanExternalID != ""
 	k8sClassSet := p.ClusterServiceClassName != ""
 	k8sPlanSet := p.ClusterServicePlanName != ""
 
-	// Can't specify both External and k8s name but must specify one.
-	if externalClassSet == k8sClassSet {
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassExternalName"), "exactly one of clusterServiceClassExternalName or clusterServiceClassName required"))
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassName"), "exactly one of clusterServiceClassExternalName or clusterServiceClassName required"))
+	// Must specify exactly one source of the class: external id, external name, k8s name.
+	if (b2i(externalClassNameSet) + b2i(externalClassIDSet) + b2i(k8sClassSet)) != 1 {
+		classSetErrMsg := "exactly one of clusterServiceClassExternalName, clusterServiceClassExternalID, or clusterServiceClassName required"
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassExternalName"), classSetErrMsg))
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassExternalID"), classSetErrMsg))
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassName"), classSetErrMsg))
 	}
-	// Can't specify both External and k8s name but must specify one.
-	if externalPlanSet == k8sPlanSet {
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanExternalName"), "exactly one of clusterServicePlanExternalName or clusterServicePlanName required"))
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanName"), "exactly one of clusterServicePlanExternalName or clusterServicePlanName required"))
+	// Must specify exactly one source of the plan: external id, external name, k8s name.
+	if (b2i(externalPlanNameSet) + b2i(externalPlanIDSet) + b2i(k8sPlanSet)) != 1 {
```

Comment:
I'm too tired to tell whether this is correct or not, so stopping here - wanted to get you preliminary feedback.